GVFS/GVFS.Virtualization/FileSystemCallbacks.cs

```diff
@@ -259,30 +263,20 @@ public bool IsReadyForExternalAcquireLockRequests(NamedPipeMessages.LockData req
         public EventMetadata GetMetadataForHeartBeat(ref EventLevel eventLevel)
         {
             EventMetadata metadata = new EventMetadata();
-            if (this.placeHolderCreationCount.Count > 0)
-            {
-                ConcurrentDictionary<string, PlaceHolderCreateCounter> collectedData = this.placeHolderCreationCount;
-                this.placeHolderCreationCount = new ConcurrentDictionary<string, PlaceHolderCreateCounter>(StringComparer.OrdinalIgnoreCase);
-
-                int count = 0;
-                foreach (KeyValuePair<string, PlaceHolderCreateCounter> processCount in
-                    collectedData.OrderByDescending((KeyValuePair<string, PlaceHolderCreateCounter> kvp) => kvp.Value.Count))
-                {
-                    ++count;
-                    if (count > 10)
-                    {
-                        break;
-                    }
-
-                    metadata.Add("ProcessName" + count, processCount.Key);
-                    metadata.Add("ProcessCount" + count, processCount.Value.Count);
-                }
-
-                eventLevel = EventLevel.Informational;
-            }
+            metadata.Add(
+                "FilePlaceholderCreation",
+                this.GetProcessInteractionData(ref this.filePlaceHolderCreationCount, ref eventLevel));

```

Comment:
Why do these need to passed by reference?

Reply:
It appears the `GetProcessInteractionData` method is setting the concurrent dictionary reference to a new instance before iterating on it, effectively resetting the dictionary each time.

Why are we doing this on the `EventMetadata.Add`?
If you do want to clear the various dictionaries I would make it explicit and obvious by calling `.Clear` on each dictionary here in _this_ method, not hidden inside the `GetProcessInteractionData` method.

Methods named "get" should probably not mutate things.

Reply:
@mjcheetham .Clear suggestion would also negate the need to pass by ref.

Reply:
```csharp
metadata.Add(
    "FilePlaceholderCreation",
    this.GetProcessInteractionData(this.filePlaceHolderCreationCount));
this.filePlaceHolderCreationCount.Clear();

// .. repeated for the other dictionaries ..

eventLevel = EventLevel.Informational;
```

Reply:
> Methods named "get" should probably not mutate things.

Agreed, this method as a whole could be better named as well (e.g. `GetAndResetHeartBeatMetadata`)

> If you do want to clear the various dictionaries I would make it explicit and obvious by calling `.Clear`

Some background info on this discussion:

The idea of using a swap here rather than `Clear` was that so if threads were adding new entries to `placeHolderCreationCount` while this method was processing the collection we'd reduce our chances of dropping anything new that was added (before calling `Clear`), and this method could take its time in processing the data.

I just tool a look and saw that `placeHolderCreationCount` is not currently marked as `volatile` though, and so even with the current code we could drop some new entries.

Reply:
Thanks for all the feedback here. I renamed some methods here and reworked how we reason about the EventLevel (which controls if we actually log this message to the mount log file).

William called out why we have the current 'swap' approach for the dictionaries and that there is still a race condition in our current approach. While the Clear approach is more explicit in our intent, I believe it would widen the window of the race condition, so I'd like to keep the swap approach as is for now.

I renamed methods here to try and make things more explicit. What does everyone think now? If it's still confusing, I can switch to `Clear`.

Reply:
I think the way you've cleaned up the code looks good, the new method names make it easier to follow what's going on.